### ISSUE TITLE
[Bug]: Error: cannot remove subnet because servers are attached to it (service_error)

### ISSUE THREAD

**Issue by vaisov** [hetznercloud/terraform-provider-hcloud#388]
When trying to destroy everything created by Terraform I get the following error:
`Error: cannot remove subnet because servers are attached to it (service_error)`

My main.tf:
```
terraform {
  backend "remote" {
    hostname     = "app.terraform.io"
    organization = "Company"

    workspaces {
      prefix     = "rancher-hetzner-"
    }
  }

  required_providers {
    hcloud = {
      source     = "hetznercloud/hcloud"
      version    = "1.26.0"
    }
  }
}

provider "hcloud" {
  token = var.hetzner_token
}

resource "hcloud_network" "lan" {
  name     = var.network_name
  ip_range = var.lan_ip_range
}

resource "hcloud_network_subnet" "lan" {
  network_id = hcloud_network.lan.id
  type = "cloud"
  network_zone = "eu-central"
  ip_range   = "192.168.0.0/24"
}

resource "hcloud_load_balancer" "load_balancer" {
  name               = "rancher"
  load_balancer_type = var.load_balancer_type
  location           = var.location
  algorithm {
    type = var.load_balancer_algorithm
  }
}

resource "hcloud_firewall" "default" {
  name = "default"
  rule {
    direction  = "in"
    protocol   = "tcp"
    port       = 22
    source_ips = toset(var.ssh_whitelist_ips)
  }

  dynamic "rule" {
    for_each = toset(var.rke_tcp_ports)
    content {
      direction  = "in"
      protocol   = "tcp"
      port       = rule.value
      source_ips = formatlist("%s/32", hcloud_server.server.*.ipv4_address)
    }
  }

  dynamic "rule" {
    for_each = toset(var.rke_udp_ports)
    content {
      direction  = "in"
      protocol   = "udp"
      port       = rule.value
      source_ips = formatlist("%s/32", hcloud_server.server.*.ipv4_address)
    }
  }
}

resource "hcloud_ssh_key" "default" {
  for_each = var.ssh_keys

  name       = each.key
  public_key = each.value
}

data "hcloud_datacenters" "ds" {
}

resource "hcloud_server" "server" {
  count       = 3
  name        = "rancher-${var.hosting_provider}-${var.environment}-${count.index+1}"
  image       = var.server_image
  server_type = var.server_type
  datacenter  = element(data.hcloud_datacenters.ds.names, count.index)
  ssh_keys    = keys(var.ssh_keys)
  labels      = { "role" : "kubernetes" }
  network {
    network_id = hcloud_network.lan.id
  }

  depends_on  = [
    hcloud_network_subnet.lan,
    hcloud_network.lan,
  ]

}

resource "hcloud_load_balancer_target" "load_balancer_target" {
  depends_on = [ hcloud_server.server ]

  type             = "label_selector"
  load_balancer_id = hcloud_load_balancer.load_balancer.id
  label_selector   = "role=kubernetes"
}
```

**Comment by fhofherr** [comment]
Hi @vaisov,

thanks for your issue and the the accompanying Terraform definition. Unfortunately this definition seems to contain additional configuration which does not seem related to the actual issue. Especially there are a lot of variables defined which can not be resolved. Additionally there is a backend for terraform cloud configured. As it currently is we cannot use the definition to reproduce your problem. Could you please provide a minimal working example that demonstrates your problem?

**Comment by vaisov** [comment]
Hi @fhofherr,

here's more simple example with variables inside. It should work on your side (just update some fields based on your environment):

```
terraform {
  required_providers {
    hcloud = {
      source     = "hetznercloud/hcloud"
      version    = "1.27.0"
    }
  }
}

provider "hcloud" {
  token = "xxxxxxx"
}

resource "hcloud_network" "lan" {
  name     = "default"
  ip_range = "192.168.0.0/24"
}

resource "hcloud_network_subnet" "lan" {
  network_id = hcloud_network.lan.id
  type = "cloud"
  network_zone = "eu-central"
  ip_range   = "192.168.0.0/24"
}

resource "hcloud_load_balancer" "load_balancer" {
  name               = "rancher"
  load_balancer_type = "lb11"
  location           = "nbg1"
  algorithm {
    type = "least_connections"
  }
}

resource "hcloud_firewall" "default" {
  name = "default"
  rule {
    direction  = "in"
    protocol   = "tcp"
    port       = 22
    source_ips = [ "1.2.3.4/32", "2.3.4.5/32" ]
  }

  dynamic "rule" {
    for_each = [ "2379", "2380", "4240", "6443", "6942", "9090", "9876", "9890", "9891", "9892", "9893", "10250", "10254" ]
    content {
      direction  = "in"
      protocol   = "tcp"
      port       = rule.value
      source_ips = formatlist("%s/32", hcloud_server.server.*.ipv4_address)
    }
  }

  dynamic "rule" {
    for_each = [ "8472" ]
    content {
      direction  = "in"
      protocol   = "udp"
      port       = rule.value
      source_ips = formatlist("%s/32", hcloud_server.server.*.ipv4_address)
    }
  }
}

resource "hcloud_ssh_key" "default" {
  for_each = { user = "ssh-rsa AAAAB3N....." }

  name       = each.key
  public_key = each.value
}

data "hcloud_datacenters" "ds" {
}

resource "hcloud_server" "server" {
  count       = 3
  name        = "rancher-hetzner-prod-${count.index+1}"
  image       = "centos-7"
  server_type = "cx31"
  datacenter  = element(data.hcloud_datacenters.ds.names, count.index)
  ssh_keys    = keys({ user = "ssh-rsa AAAAB3N....." })
  labels      = { "role" : "kubernetes" }
  network {
    network_id = hcloud_network.lan.id
  }

  depends_on  = [
    hcloud_network_subnet.lan,
    hcloud_network.lan,
  ]

}

resource "hcloud_load_balancer_target" "load_balancer_target" {
  depends_on = [ hcloud_server.server ]

  type             = "label_selector"
  load_balancer_id = hcloud_load_balancer.load_balancer.id
  label_selector   = "role=kubernetes"
}
```

**Comment by LKaemmerling** [comment]
Hey @vaisov,

can you try it with the latest version (v1.27.1?)

**Comment by vaisov** [comment]
The issue still persists even with `v1.27.2`

**Comment by LKaemmerling** [comment]
Hey @vaisov,

would it be possible that you send us a log of the `terraform apply` when you delete the resources? You can control the logging via ENVs.

```
export TF_LOG=debug
export TF_LOG_PATH=debug.txt
terraform apply
```

Please send this log file to lukas.kaemmerling@hetzner-cloud.de

**Comment by vaisov** [comment]
Log file sent

**Comment by vaisov** [comment]
Update from my side, last time I ran the destroy from CLI instead of automated destroy from Terraform Cloud and it worked, but only when I ran it for the second time.

First time `terraform destroy`:
```
module.rke.hcloud_network_subnet.lan: Destroying... [id=1170706-192.168.0.0/24]
module.rke.hcloud_network_subnet.lan: Still destroying... [id=1170706-192.168.0.0/24, 10s elapsed]
╷
│ Error: cannot remove subnet because servers are attached to it (service_error)
│
│
╵
```

Second time:
```
module.rke.hcloud_network_subnet.lan: Destroying... [id=1170706-192.168.0.0/24]
module.rke.hcloud_network_subnet.lan: Destruction complete after 1s
module.rke.hcloud_network.lan: Destroying... [id=1170706]
module.rke.hcloud_network.lan: Destruction complete after 0s

Apply complete! Resources: 0 added, 0 changed, 2 destroyed.
```

**Comment by thelittlebug** [comment]
i tried it via the hetzner webinterface:
i'm not able to remove the subnet.
but i'm able to remove the network.

so it seems to be an error at the api.

edit: it takes some time (random for me, something between 1 and 5 minutes) until i'm able to remove the subnet via the webinterface.

**Comment by lflaimer** [comment]
I am facing the same issue, any update on a fix ?

**Comment by LKaemmerling** [comment]
We will have a solution in the next provider release. It was merged with #441
